plugins/experimental/carp/Common.cc

```diff
@@ -0,0 +1,364 @@
+      unsigned int iLPort = 0;
+      sscanf(vLocalParts[1].c_str(), "%x", &iLPort);
+      TSDebug(DEBUG_TAG_INIT, "Found listening port %d", iLPort);
+      if (iLPort == iPort) {
+        string sLabel = string("socket:[" + vParts[9] + "]");
+        // find inode in   /proc/{pid}/fd/
+        string sDir = "/proc/" + string(sPid) + "/fd";
+
+        DIR *dp;
+        struct dirent *dirp;
+        if ((dp = opendir(sDir.c_str())) == NULL) {
+          TSDebug(DEBUG_TAG_INIT, "Failed to open directory %s, %s", sDir.c_str(), strerror(errno));
+          continue;
+        }
+
+        while ((dirp = readdir(dp)) != NULL && !bParseDone) {
+          //TSDebug(DEBUG_TAG_INIT, "File in  %s = %s inode=%d", sDir.c_str(),dirp->d_name,dirp->d_ino);
+          string sPath = sDir + "/" + dirp->d_name;
+          char sBuf[256];
+          ssize_t iCount = readlink(sPath.c_str(), sBuf, sizeof (sBuf));
+          if (iCount > 0) {
+            string sLink(sBuf, iCount);
+            if (sLink.compare(sLabel) == 0) {
+              TSDebug(DEBUG_TAG_INIT, "Found that port %d is opened for listening by pid %s", iLPort, sPid.c_str());
+              bParseDone = true;
+              bMatch = true;
+            }
+          }
+        }
+        closedir(dp);
+      }
+    }
+    /*    
+        for(int i=0;i<vParts.size();i++)
+        {
+          TSDebug(DEBUG_TAG_INIT, "part[%d]='%s'", i, vParts[i].c_str());
+        }
+     */
+  }
+
+  TSfclose(file);
+
+  return bMatch;
+}
+
+/************************************************************************/
+bool 
+isPortSelf(unsigned int iPort)
+{
+  pid_t pid = getpid();
+  char sPid[10];
+  bool bMatch = false;
+
+  string sFileName; 
+  // open file @ /proc/{pid}/net/tcp
+
+  
+  sprintf(sPid,"%d",pid);
+  
+  // look for IPv4 listener first
+  sFileName = "/proc/" + string(sPid) + "/net/tcp";
+  bMatch = scanProcFileForPort(sFileName,sPid, iPort);
+  if(!bMatch) { // did not find IPv4 listener, check for IPv6
+    sFileName = "/proc/" + string(sPid) + "/net/tcp6";
+    bMatch = scanProcFileForPort(sFileName,sPid, iPort);
+  }
+  return bMatch;
+}
+
+/************************************************************************/
+struct hostent * 
+getHostIp(string hName, struct hostent *h, char *buf, int buflen)
+{
+    int res, err;
+    struct hostent *hp = NULL;
+
+    res = gethostbyname_r(hName.c_str(), h, buf, buflen, &hp, &err);
+    if ((res == 0) && (hp != NULL)) {
+      return hp;
+    } else {
+      TSDebug(DEBUG_TAG_INIT, "gethostbyname_r failed for %s.  Error=%s", hName.c_str(), strerror(err));
+    }
+    return NULL;
+}
+
+          
+/************************************************************************/
+/*
+ * *** WARNING *** You will need to run the carp plugin with traffic_manager for it to detect
+ * itself and to forward directly to the origin.  It will not work by running traffic_server directly!
+ */
```

Comment:
Why? Because `traffic_manager` opens the ports and so only if it is running will the port scan work? In that case, why not wait until `TS_LIFECYCLE_PORTS_READY_HOOK`?